plugins/flatpak/gs-flatpak.c

```diff
@@ -3678,6 +3678,75 @@ app_has_local_source (GsApp *app)
 		(url != NULL && g_str_has_prefix (url, "file://"));
 }
 
+static guint64
+get_installation_dir_free_space (GsFlatpak *self)
+{
+	g_autoptr (GFile) installation_dir = NULL;
+	g_autoptr (GFileInfo) info = NULL;
+	g_autoptr (GError) local_error = NULL;
+	guint64 free_space = 0;
+
+	installation_dir = flatpak_installation_get_path (self->installation);
+
+	info = g_file_query_filesystem_info (installation_dir,
+					     G_FILE_ATTRIBUTE_FILESYSTEM_FREE,
+					     NULL,
+					     &local_error);
+	if (info == NULL)
+		g_warning ("Failed to query filesystem info: %s", local_error->message);
+
+	free_space = g_file_info_get_attribute_uint64 (info, G_FILE_ATTRIBUTE_FILESYSTEM_FREE);
+	return free_space;
+}
+
+static gboolean
+gs_flatpak_has_space_to_install (GsFlatpak *self, GsApp *app, GsAppList *list)
+{
+	guint64 free_space = 0;
+	guint64 space_required = 0;
+
+	space_required = gs_app_get_size_download (app);
+	if (space_required == GS_APP_SIZE_UNKNOWABLE)
+		g_warning ("Failed to query download size: %s", gs_app_get_unique_id (app));
+
+	space_required = space_required + (500 * 1024 * 1024); /* min-free-space-size = 500MB for Endless */
+	free_space = get_installation_dir_free_space (self);
+
+	if (free_space < space_required)
+		return FALSE;
+
+	return TRUE;
+}
+
+static gboolean
+gs_flatpak_has_space_to_update (GsFlatpak *self, GsApp *app, GsAppList *list)
+{
+	guint64 free_space = 0;
+	guint64 space_required = 0;
+
+	/* If an app has an updatable runtime, the runtime is not updated
+	 * as a part of app update.
+	 *
+	 * Having said that, runtime updates and their related refs (see T22716)
+	 * are done together and below heuristics might not be good for if GsApp
+	 * app is a runtime.
+	 */
```

Comment:
What does this mean? It doesn't matter that the runtime is not updated together with the app for this task.
But you should check this code for runtime updates too. I think it will work fine, but in any case this comment is confusing and shouldn't be necessary (please do test both apps + runtimes).

Reply:
This is something I need some discussion on. The heuristics is bad if GsApp is a Runtime. For e.g. GNOME runtime is around ~1.6GB when installed, so if there is a runtime update pending; according to this heuristics it will dictate to have `(1.6GB * 2 + 500MB)` space required for update to happen.

Reply:
Yeah, the assumption I detailed in the ticket was that we should only use this heuristic for app upgrades, on the basis that most files are likely to be changed. For runtimes, that's (hopefully!) far less likely. If we could get the delta size, it would be best, but for the moment we should just leave it as checking the min space.